### ISSUE TITLE
"exit 123" can return to "father's father" script,is this by design?

### ISSUE THREAD

**Issue by chuanjiao10** [PowerShell/PowerShell#10629]
hi all：

https://github.com/PowerShell/PowerShell-RFC/blob/e622135ce5ced14517543739802a157eaa92aa40/1-Draft/RFCNNNN-RFC-Native-Command-Exit-Errors.md

https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_language_keywords?view=powershell-6#exit

i have seen up url，can not find "exit 123" to father's father script should be.

i looking in powershell repo too,can not find message.

Steps to reproduce
------------------
#a.ps1 call b.ps1 ,b.ps1 call c.ps1 
#Not related to "&", "."

```
#a.ps1
& 'b.ps1'
$LASTEXITCODE
```

```
#b.ps1
& 'c.ps1'
$LASTEXITCODE
```

```
#c.ps1
exit 123
```

linux + bash return
-----------------
123
0

Actual behavior
---------------
123
123

my question
------------------
Q:
"exit 123" can return to father's father script,is this by design?
if true, I hope I can update the instructions in the manual.

by the way,i tested this in cmd   .bat
------------------
i foud confusion!!! Execute the bat script twice, the return value is different!

Environment data
----------------
win10 + powershell 6.2.3
win10 + powershell 5.1

reporter
----------------
By the chinese powershell study & research group. QQ group No. is 183173532

**Comment by mklement0** [comment]
Unless you explicitly use `exit` in PowerShell (or call an external program), the - global - `$LASTEXITCODE` is not updated. (Using just `exit` rather than `exit <n>` sets the value to `0`).

Therefore, it is the most recent `exit`  statement or external-program call (whichever came last) that sets the `$LASTEXITCODE` value, in whatever scope they were executed, and that value remains in effect until another such call is made.

Exit codes are a PowerShell-independent mechanism for signaling status between _processes_.
 
PowerShell's support for it is designed for integration with the _outside_ world, notably when the CLI is used, it is not designed to be used _between PowerShell scripts_.

When using the PowerShell CLI:

* With `-File`, the target script's `exit` call is honored as-is in terms of the resulting process exit code.
* With `-Command` it is the last statement in the command string whose success / failure status - as reflected in `$?` - is reported as `1` (failure) or `0`; however, you can use `exit <n>` as the last statement to control the exit code explicitly (e.g., 
`pwsh -nop -c 'Get-Item /nosuch; if (-not $?) { exit 5 }`')

**Comment by chuanjiao10** [comment]
> Therefore, it is the most recent exit statement or external-program call (whichever came last) that sets the $LASTEXITCODE value, in whatever scope exit / an external program was called, and that value remains in effect until another such call is made.
@mklement0 
Welcome feedback,These sentences are very vague.

The actual situation is:
1 "exit 123" can penetrate all script calls.     a.ps1--->b.ps1 --->c.ps1 ,exit 123 can back to a.ps1
2 pwsh -file The parent script .ps1 , "exit 123" in the parent script, can penetrate the process.
```
& "C:\Program Files\PowerShell\6\pwsh.exe" -file 'a:\pscode\TEMP_2019\temp210\parent script.ps1'
"$LASTEXITCODE in another process"
```
2 pwsh -file The parent script .ps1 , the parent script is called from the script, and the "exit 123" in the child cannot penetrate the process.

My point is:
"Subscript" can "exit 1", "parent script" can also, This is very common.
now exit can penetrate, then "grandfather script" does not know at all, who is ("subscript" or "parent script") return, I suggest blocking exit Penetrate the script.

@SteveL-MSFT  @vexx32  ：
have a look，thanks

**Comment by mklement0** [comment]
@chuanjiao10: 

Re 1: 
That is explained by my previous comment.

Re 2: 

Indeed, only the script _directly_ passed to `-File` can set the process exit code for the CLI invocation, `exit` calls in nested script calls are ignored, as are the exit codes reported by external programs called by the script  - I've updated the previous comment to make that clear.

As for your point:

* In _CLI use_, `exit` from a _nested_ call doesn't propagate ("penetrate"), as discussed above - with `-File`, the outside world sees just the top-level script's exit code, which is implicitly `0` (if the script isn't exited with an `exit` call) or whatever `<n>` is in an `exit <n>` call.

* _Inside_ a PowerShell session, exit calls do propagate in the sense that `$LASTEXITCODE` reflects the last `exit` call from a nested script call (however many nesting levels are involved) - but the outside world won't see that value, unless the top-level script explicitly relays it with `exit $LASTEXITCODE`.

Yes, you won't be able to tell _which_ among nested script calls actually caused `$LASTEXITCODE` to be set. If that is a concern, you must ensure that _all_ of your scripts exit with an `exit` statement.

Again, exit codes aren't designed to communicate status information _inside_ of PowerShell, between scripts that call each other; they're a _process-level_ concept whose primary purpose is to communicate with the _outside_ world.

You could argue that `exit` should always exit the _entire process_, as is the case in scripts for POSIX-like shells, in which case the subtleties of the current behavior would not arise - but it's clearly too late to change that.

There's also the awkwardness that calling `exit` inside a _function_ indeed _does_ exit the process - in whatever scope it is called.

Let me attempt to provide pragmatic guidance:

* Whenever you author a script that you (also) expect to be called via the CLI's `-File` parameter and you deliberately want to set a process exit code, make sure that your script exits with an `exit` statement (without `exit`, the process exit code will be `0`, unless a script-terminating error occurs, in which case it is `1`).

* Similarly, when you use the CLI's `-Command` parameter and you deliberately want to set a process exit code, end your command string with an explicit `exit` statement (without it, the process exit code will be `0` if `$?` happens to contain `$true` after executing the last statement, and `1` if it contains `$false` - since `$?` is an unreliable indicator of success vs. failure, you shouldn't rely on it; as with `-File`, a script-terminating error results in exit code `1`).

* When calling other scripts _from within PowerShell_, do not rely on `$LASTEXITCODE`, unless you happen to know that a given script deliberately sets it; otherwise:
  * use `$LASTEXITCODE` only to inspect the process exit code from the most recent call to an _external program_,
  * rely on PowerShell's native error-reporting and -handling features to communicate success between scripts and functions.

**Comment by chuanjiao10** [comment]
> exit codes aren't designed to communicate status information inside of PowerShell, between scripts that call each other
I am afraid I can’t agree.
The voice from the Chinese forum: "exit 0 In every powershell script, this is a cancer of powershell. The default value of exit for other languages is 0, no need to do this"

**Comment by mklement0** [comment]
Well, exit codes are _process_ exit codes, and when PowerShell scripts call each other, there are no processes that exit - it all happens _in-process_.

However, I agree that the current intra-PowerShell behavior with respect to `exit` is not really useful, and that it would be more useful to mimic the behavior that the top-level script passed to `-File` exhibits, which would mean:

Set `$LASTEXITCODE` to `0` whenever a script exits without an explicit `exit` statement.

However, that would be a breaking change, because `$LASTEXITCODE` would no longer reflect an external program's exit code across nested script calls (that don't use `exit`).

Conceivably, there could be code out there that calls an external program in `b.ps1`, and relies on that program's exit code to still be reflected in `$LASTEXITCODE` in `a.ps1` after calling `b.ps1`; with the above change, `b.ps1` exiting would reset `$LASTEXITCODE`.

In short: `$LASTEXITCODE` would then only reflect the most recently executed external program's exit code _until any script exits_.
